processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -180,26 +231,27 @@ public int getNumRows()
 
     private List<ByteBuffer> updateValues(
```

Comment:
The GroupByEngine v1 used recursion with sublists as well, so I maintained that behavior, and benchmarks numbers between master and this patch didn't show performance degradation for GroupBy. 

Given that GroupByEngine v1 will eventually be deprecated by the newer V2 engine (which uses a stack for the functionality in question), I'm leaning towards keeping the recursion here

There are other issues with GroupByEngine V1 as well (e.g, see testSubqueryWithOuterMaxOnDiskStorageContextOverride() in GroupByRunnerTest)

I made that "use index instead of creating sublist" change as suggested though.